### PR TITLE
[PlayStation] Web Inspector: add memory usage of page in memory timeline

### PR DESCRIPTION
**Pull request by haruhisa-shin**
#### d93e32a105054e1b9cf55b8c3ae291775019da2b
<pre>
[PlayStation] Web Inspector: add memory usage of page in memory timeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=255642">https://bugs.webkit.org/show_bug.cgi?id=255642</a>

Reviewed by Basuke Suzuki.

Currently, the inspector backend of playstation only aggregates the size of GCHeap.
This change adds the usage of page category to memory details in timeline tab.

* Source/WebCore/PlatformPlayStation.cmake:
* Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp:
(WebCore::ResourceUsageThread::platformCollectMemoryData):

Canonical link: <a href="https://commits.webkit.org/263405@main">https://commits.webkit.org/263405@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/53b6b94d4f38621490ba37bdc1643ca2af855d5e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6015 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4604 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/4935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4698 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4059 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6021 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4049 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4047 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4121 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5654 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4504 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4048 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1106 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8075 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4400 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 1]
```diff
@@ -55,17 +57,31 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+    showmap::Result<256> result;
+    result.collect();
+    if (auto* entry = result.entry("SceNKFastMalloc"))
+        categories[MemoryCategory::bmalloc].dirtySize = entry->effectiveRss();
+
+    // TODO: collect dirty size of "MemoryCategory::Images" and "MemoryCategory::Layers"
+    if (data.totalDirtySize > categories[MemoryCategory::bmalloc].dirtySize)
+        categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - categories[MemoryCategory::bmalloc].dirtySize;
+
+    auto& dirtySize = categories[isFastMallocEnabled() ? MemoryCategory::bmalloc : MemoryCategory::Other].dirtySize;
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Relationship between `categories[MemoryCategory::bmalloc]` and `categories[MemoryCategory::Other]` is a bit unclear.

Can it be more clear if their shape is like this?
```
if (isFastMallocEnabled()) {
   ...
} else {
   ...
}
```

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 2]
```diff
@@ -55,17 +57,31 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
 
-    data.categories[MemoryCategory::GCHeap].dirtySize = currentGCHeapCapacity;
-    data.categories[MemoryCategory::GCOwned].dirtySize = currentGCOwnedExtra - currentGCOwnedExternal;
-    data.categories[MemoryCategory::GCOwned].externalSize = currentGCOwnedExternal;
+    categories[MemoryCategory::GCHeap].dirtySize = currentGCHeapCapacity;
+    categories[MemoryCategory::GCOwned].dirtySize = currentGCOwnedExtra - currentGCOwnedExternal;
+    categories[MemoryCategory::GCOwned].externalSize = currentGCOwnedExternal;
+
+    showmap::Result<256> result;
+    result.collect();
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp:75]: Showmap's `collect()` requires several kernel calls and not a cheap operation. You don't need to check that if `isFastMallocEnabled()` is `false`.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp (hunk 1 above)
Comment (haruhisa-shin) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Then the `auto&` cannot be used, so I made a similar code based on cocoa port.
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/page/cocoa/ResourceUsageThreadCocoa.mm#L262

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp (hunk 2 above)
Comment (haruhisa-shin) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp:75]: Sure. Fixed by https://github.com/WebKit/WebKit/pull/12902/commits/0911d72bc2715f5bd58169779170438e137aeaf7

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 3]
```diff
@@ -55,17 +57,33 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+
+    if (isFastMallocEnabled()) {
+        showmap::Result<256> result;
+        result.collect();
+        if (auto* entry = result.entry("SceNKFastMalloc"))
+            categories[MemoryCategory::bmalloc].dirtySize = entry->effectiveRss();
+
+        // TODO: collect dirty size of "MemoryCategory::Images" and "MemoryCategory::Layers"
+        if (data.totalDirtySize > categories[MemoryCategory::bmalloc].dirtySize)
+            categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - categories[MemoryCategory::bmalloc].dirtySize;
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: This is wrong. When not isFastMallocEnabled(), categories[MemoryCategory::bmalloc].dirtySize is zero so it must be same with data.totalDirtySize

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 4]
```diff
@@ -55,17 +57,33 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+    }
+
+    auto& mallocBucket = isFastMallocEnabled()? data.categories[MemoryCategory::bmalloc]: data.categories[MemoryCategory::Other];
+    mallocBucket.dirtySize -= currentGCHeapCapacity;
+    mallocBucket.dirtySize -= std::min(mallocBucket.dirtySize, categories[MemoryCategory::GCOwned].dirtySize);
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Please put this logic inside if and else  even if some logic was duplicated to simplify overall code.

**Review by basuke — COMMENTED** [review]
Still code looks complicated. Please divide the logic if and else (except currentGCHeapXXX part).

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp (hunk 3 above)
Comment (haruhisa-shin) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Updated.
I put `categories[MemoryCategory::Other].dirtySize = data.totalDirtySize;` in the "else" side.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 5]
```diff
@@ -55,17 +57,40 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+
+    if (isFastMallocEnabled()) {
+        auto& mallocBucket = categories[MemoryCategory::bmalloc];
+
+        showmap::Result<256> result;
+        result.collect();
+        if (auto* entry = result.entry("SceNKFastMalloc"))
+            mallocBucket.dirtySize = entry->effectiveRss();
+
+        // TODO: collect dirty size of "MemoryCategory::Images" and "MemoryCategory::Layers"
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: This comment is for both cases. Move out from if and else.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 6]
```diff
@@ -55,17 +57,40 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+            categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - mallocBucket.dirtySize;
+
+        mallocBucket.dirtySize -= currentGCHeapCapacity;
+        mallocBucket.dirtySize -= std::min(mallocBucket.dirtySize, categories[MemoryCategory::GCOwned].dirtySize);
+    } else {
+        auto& mallocBucket = categories[MemoryCategory::Other];
+        mallocBucket.dirtySize = data.totalDirtySize;
+
+        mallocBucket.dirtySize -= currentGCHeapCapacity;
+        mallocBucket.dirtySize -= std::min(mallocBucket.dirtySize, categories[MemoryCategory::GCOwned].dirtySize);
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Else part of this four lines can be written in one line:
```
categories[MemoryCategory::Other] = data.totalDirtySize - std::min(data.totalDirtySize, currentGCHeapCapacity + currentGCOwnedExtra - currentGCOwnedExternal);
```
or two lines for readability:
```
auto currentGCDirtySize = categories[MemoryCategory::GCHeap].dirtySize + categories[MemoryCategory::GCOwned].dirtySize;
categories[MemoryCategory::Other] = data.totalDirtySize - std::min(data.totalDirtySize, currentGCDirtySize);
```

Comment (basuke) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: And definition of `currentGCDirtySize`  can be moved just after definition of `currentGCOwnedExternal` using original values:
```
auto currentGCDirtySize = currentGCHeapCapacity + currentGCOwnedExtra - currentGCOwnedExternal;
```
so the else part is again one line :)

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 7]
```diff
@@ -55,17 +57,40 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+    categories[MemoryCategory::GCHeap].dirtySize = currentGCHeapCapacity;
+    categories[MemoryCategory::GCOwned].dirtySize = currentGCOwnedExtra - currentGCOwnedExternal;
+    categories[MemoryCategory::GCOwned].externalSize = currentGCOwnedExternal;
+
+    if (isFastMallocEnabled()) {
+        auto& mallocBucket = categories[MemoryCategory::bmalloc];
+
+        showmap::Result<256> result;
+        result.collect();
+        if (auto* entry = result.entry("SceNKFastMalloc"))
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: If this return nullptr, `mallocBucket.dirtySize` might be overflowed. Actually we can assume `entry` is always non null. So remove `if` and put `RELEASE_ASSERT(entry);` to ensure the logic is correct.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 8]
```diff
@@ -55,17 +57,40 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+            mallocBucket.dirtySize = entry->effectiveRss();
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Stop using `mallocBucket` in the block. assign auto variable `auto rss = entry->effectiveRss()` here. Then (continue to [*1])

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 9]
```diff
@@ -55,17 +57,40 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+
+        // TODO: collect dirty size of "MemoryCategory::Images" and "MemoryCategory::Layers"
+        if (data.totalDirtySize > mallocBucket.dirtySize)
+            categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - mallocBucket.dirtySize;
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: [*1] use `rss` to calculate `categories[MemoryCategory::Other].dirtySize` and replace following two lines using same technique with else part (see below).

**Review by basuke — CHANGES_REQUESTED** [review]
Sorry, but keep changing to make the logic easy to understand.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 10]
```diff
@@ -55,17 +56,33 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+
+    // TODO: collect dirty size of "MemoryCategory::Images" and "MemoryCategory::Layers"
+
+    showmap::Result<256> result;
+    result.collect();
+    data.totalDirtySize = result.effectiveRss();
+
+    if (auto* entry = result.entry("SceNKFastMalloc")) {
+        auto rss = entry->effectiveRss();
+        if (data.totalDirtySize > rss)
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: This is always true. Change it to RELEASE_ASSERT

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp [hunk 11]
```diff
@@ -55,17 +56,33 @@ void ResourceUsageThread::platformCollectCPUData(JSC::VM*, ResourceUsageData& da
...
+            categories[MemoryCategory::Other].dirtySize = data.totalDirtySize - rss;
+
+        categories[MemoryCategory::bmalloc] = rss - std::min(rss, currentGCDirtySize);
```
Comment (basuke) [inline Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: This should come first than assigning to `Other`. Then `Other` = `totalDirtySize` - `bmalloc`.

**Review by basuke — CHANGES_REQUESTED** [review]
Much simpler. We are going to right direction. One more tiny change, then it should be okay.

File: Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp (hunk 11 above)
Comment (haruhisa-shin) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Then, the `currentGCDirtySize` would be included in `Other`. Is that correct?

Comment (haruhisa-shin) [reply Source/WebCore/page/playstation/ResourceUsageThreadPlayStation.cpp]: Sorry, but I did not change this order.
If there is still a mistake, please point it out again.

**Review by basuke — APPROVED** [review]
LGTM! Thanks for fixing. The code looks much better than before.

**Pull request merged**